iati/core/rulesets.py

```diff
@@ -233,22 +253,72 @@ def _ruleset_schema_section(self):
         ruleset_schema = iati.core.default.ruleset_schema()
         partial_schema = ruleset_schema['patternProperties']['.+']['properties'][self.name]['properties']['cases']['items']  # pylint: disable=E1101
         # make all attributes other than 'condition' in the partial schema required
-        partial_schema['required'] = self._required_case_attributes(partial_schema)
+        partial_schema['required'] = self._case_attributes(partial_schema)
         # ensure that the 'paths' array is not empty
         if 'paths' in partial_schema['properties'].keys():
             partial_schema['properties']['paths']['minItems'] = 1
 
         return partial_schema
 
+    def _find_context_elements(self, dataset):
+        """Find the specific elements in context for the Rule.
+
+        Args:
+            dataset (iati.core.Dataset): The Dataset to be chacked for validity against the Rule.
+
+        Returns:
+            list: A list of all elements found for the given context.
+
+        """
+        return dataset.xml_tree.xpath(self.context)
+
+    def _extract_text_from_element_or_attribute(self, xpath_results):
+        """Return a list of strings regardless of whether XPath result is an attribute or an element.
+
+        Args:
+            xpath_results (list): Raw XPath query results.
+
+        Returns:
+            A list of strings.
```

Comment:
`A list of strings` -> `list of str: [something about what the strings in the list are]`

Reply:
done